hetmech/dwpc.py

```diff
@@ -0,0 +1,108 @@
+import functools
+import operator
+
+import numpy
+
+from .matrix import metaedge_to_adjacency_matrix, normalize, copy_array
+
+
+def remove_diag(mat):
+    """Set the main diagonal of a square matrix to zeros."""
+    assert len(set(mat.shape)) == 1  # must be square
+    return mat - numpy.diag(mat.diagonal())
+
+
+def degree_weight(matrix, damping, copy=True):
+    """Normalize an adjacency matrix by the in and out degree."""
+    matrix = copy_array(matrix, copy)
+    row_sums = numpy.array(matrix.sum(axis=1)).flatten()
+    column_sums = numpy.array(matrix.sum(axis=0)).flatten()
+    matrix = normalize(matrix, row_sums, 'rows', damping)
+    matrix = normalize(matrix, column_sums, 'columns', damping)
+
+    return matrix
+
+
+def dwpc_no_repeats(graph, metapath, damping=0.5):
+    assert len(set(metapath.edges)) == len(metapath)
+
+    parts = list()
+    for edge in metapath:
+        rows, cols, adj = metaedge_to_adjacency_matrix(
+            graph, edge, dtype=numpy.float64, sparse_threshold=0)
+        adj = degree_weight(adj, damping)
+
+        if edge == metapath[0]:
+            row_names = rows
+        if edge == metapath[-1]:
```

Comment:
This is nice and explicit, but you can always just returns cols, which will be set from the last iteration of the loop. Your choice, since sometimes being clear is more important than being brief.